op-e2e/sgt/sgt_test.go

```diff
@@ -0,0 +1,307 @@
+			depositSgtValue: big.NewInt(10000000000000),
+			depositL2Value:  big.NewInt(10000000000000),
+			txValue:         big.NewInt(0),
+			action:          sgtTxSuccess,
+		},
+		{
+			name:            "PartialSGTGasPaymentSuccess",
+			depositSgtValue: big.NewInt(1000),
+			depositL2Value:  big.NewInt(10000000000000),
+			txValue:         big.NewInt(0),
+			action:          sgtTxSuccess,
+		},
+		{
+			name:            "FullSGTGasPaymentAndNonZeroTxValueWithSufficientNativeBalanceSuccess",
+			depositSgtValue: big.NewInt(10000000000000),
+			depositL2Value:  big.NewInt(10000000000000),
+			txValue:         big.NewInt(10000),
+			action:          sgtTxSuccess,
+		},
+		{
+			name:            "PartialSGTGasPaymentAndNonZeroTxValueWithSufficientNativeBalanceSuccess",
+			depositSgtValue: big.NewInt(1000),
+			depositL2Value:  big.NewInt(10000000000000),
+			txValue:         big.NewInt(10000),
+			action:          sgtTxSuccess,
+		},
+		{
+			name:            "InsufficientGasPaymentFail",
+			depositSgtValue: big.NewInt(10000),
+			depositL2Value:  big.NewInt(10000),
+			txValue:         big.NewInt(0),
+			expectedErr:     errorInsufficientBalance,
+			action:          sgtTxFail,
+		},
+		{
+			name:            "FullSGTGasPaymentAndNonZeroTxValueWithInsufficientNativeBalanceFail",
+			depositSgtValue: big.NewInt(10000000000000),
+			depositL2Value:  big.NewInt(10000),
+			txValue:         big.NewInt(10000000000000),
+			expectedErr:     errorInsufficientBalance,
+			action:          sgtTxFail,
+		},
+		{
+			name:            "PartialSGTGasPaymentAndNonZeroTxValueWithInsufficientNativeBalanceFail",
+			depositSgtValue: big.NewInt(10000),
+			depositL2Value:  big.NewInt(10000),
+			txValue:         big.NewInt(10000000000000),
+			expectedErr:     errorInsufficientBalance,
+			action:          sgtTxFail,
+		},
+	}
+
+	for index, tCase := range tests {
+		t.Run(tCase.name, func(t *testing.T) {
+			b, err := tCase.action(ctx, t, int64(index), tCase.depositSgtValue, tCase.depositL2Value, tCase.txValue, sgt)
+			require.EqualValues(t, tCase.expectedErr, err)
+			if tCase.depositSgtValue.Cmp(b.gasCost) >= 0 {
+				if tCase.depositL2Value.Cmp(tCase.txValue) >= 0 {
+					require.Equal(t, new(big.Int).Sub(tCase.depositSgtValue, b.gasCost).Cmp(b.sgtBalance), 0)
+					require.Equal(t, tCase.depositL2Value.Sub(tCase.depositL2Value, tCase.txValue).Cmp(b.l2Balance), 0)
+				}
+			} else {
+				balance := tCase.depositSgtValue.Add(tCase.depositSgtValue, tCase.depositL2Value)
+				if balance.Cmp(b.gasCost) >= 0 {
+					require.Equal(t, int64(0), b.sgtBalance.Int64())
+					require.Equal(t, balance.Sub(balance, b.gasCost).Cmp(b.l2Balance.Add(b.l2Balance, tCase.txValue)), 0)
+				}
+			}
+			if b.vaultBalanceDiff != nil {
+				// todo: how to check vault's balance diff
+			}
+		})
+	}
+
+}
+
+func sgtDeposit(ctx context.Context, t *testing.T, index int64, sgtValue *big.Int, l2Value *big.Int, txValue *big.Int, sgt *SgtHelper) (*BalanceCheck, error) {
```

Comment:
Please add a comment to each action about what it's intended to test.

Reply:
done in https://github.com/ethstorage/optimism/pull/125/commits/2823c0fdf12e0329ed54e4bd3610d62822394a4b